source/iterator/snapshot.go

```diff
@@ -0,0 +1,322 @@
+// Copyright © 2023 Meroxa, Inc. & Yalantis
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-neo4j/config"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/db"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/dbtype"
+)
+
+const (
+	// all Cypher queries used by the [Snapshot] are listed below in the format of Go fmt.
+	getNodeMaxPropertyQueryTemplate         = "MATCH (obj:%s) RETURN obj.%s as %s ORDER BY obj.%s DESC LIMIT 1"
+	getRelationshipMaxPropertyQueryTemplate = "MATCH ()-[obj:%s]-() RETURN obj.%s as %s ORDER BY obj.%s DESC LIMIT 1"
+	getNodesQueryTemplate                   = "MATCH (obj:%s) %s RETURN obj LIMIT %d"
+	getRelationshipsQueryTemplate           = "MATCH (src)-[obj:%s]->(trgt) %s RETURN obj, src, trgt LIMIT %d"
+	opmvLTEWhereClause                      = "obj.%s <= $opmv"
+	opvGTWhereClause                        = "obj.%s > $opv"
+
+	// some helpers for Cypher queries.
+	orderingPropertyMaxValueFieldName = "opmv"
+	orderingPropertyValueFieldName    = "opv"
+	objPlaceholder                    = "obj"
+	srcPlaceholder                    = "src"
+	trgtPlaceholder                   = "trgt"
+
+	// relationship payload-specific fields.
+	sourceNodeField = "sourceNode"
+	targetNodeField = "targetNode"
+)
+
+// Snapshot implements a snapshot logic for the connector.
+type Snapshot struct {
+	driver                   neo4j.DriverWithContext
+	orderingProperty         string
+	keyProperties            []string
+	orderingPropertyMaxValue any
+	entityType               config.EntityType
+	entityLabels             string
+	batchSize                int
+	databaseName             string
+	position                 *position
+	// use a channel here, as it's easier to work with it as a queue
```

Comment:
please add short description about this channel

Reply:
I've added a description to this field.